provider/infoblox.go

```
@@ -0,0 +1,389 @@
+}
+
+func (p *InfobloxProvider) newObject(e *endpoint.Endpoint, getRef bool) (obj ibclient.IBObject, ref string, err error) {
+	switch e.RecordType {
+	case endpoint.RecordTypeA:
+		var records []ibclient.RecordA
+		obj = ibclient.NewRecordA(
+			ibclient.RecordA{
+				Name:     e.DNSName,
+				Ipv4Addr: e.Target,
+			},
+		)
+		if getRef {
+			p.client.GetObject(obj, "", &records)
+			if len(records) == 1 {
+				ref = records[0].Ref
+			}
+		}
+		return obj, ref, nil
+	case endpoint.RecordTypeCNAME:
+		var records []ibclient.RecordCNAME
+		obj = ibclient.NewRecordCNAME(
+			ibclient.RecordCNAME{
+				Name:      e.DNSName,
+				Canonical: e.Target,
+			},
+		)
+		if getRef {
+			p.client.GetObject(obj, "", &records)
+			if len(records) == 1 {
+				ref = records[0].Ref
+			}
+		}
+		return obj, ref, nil
+	case endpoint.RecordTypeTXT:
+		var records []ibclient.RecordTXT
+		if target, err := strconv.Unquote(e.Target); err == nil && !strings.Contains(e.Target, " ") {
+			e.Target = target
+		}
+		obj = ibclient.NewRecordTXT(
+			ibclient.RecordTXT{
+				Name: e.DNSName,
+				Text: e.Target,
+			},
+		)
+		if getRef {
+			p.client.GetObject(obj, "", &records)
+			if len(records) == 1 {
+				ref = records[0].Ref
+			}
+		}
+		return obj, ref, nil
+	default:
+		return nil, "", fmt.Errorf("unsupported record type '%s'", e.RecordType)
+	}
+}
+
+func (p *InfobloxProvider) createRecords(created infobloxChangeMap) {
+	for zone, endpoints := range created {
+		for _, endpoint := range endpoints {
+			if p.dryRun {
+				logrus.Infof(
+					"Would create %s record named '%s' to '%s' for Infoblox DNS zone '%s'.",
+					endpoint.RecordType,
+					endpoint.DNSName,
+					endpoint.Target,
+					zone,
+				)
+				continue
+			}
+
+			logrus.Infof(
+				"Creating %s record named '%s' to '%s' for Infoblox DNS zone '%s'.",
+				endpoint.RecordType,
+				endpoint.DNSName,
+				endpoint.Target,
+				zone,
+			)
+
+			obj, _, err := p.newObject(endpoint, false)
+			if err == nil {
```

i think more natural would be to "fail early" in this case, i.e. 

```go
if err != nil { 
      // return error 
}
_, err = p.client.CreateObject(obj)
```

---

same for  `update` and `delete` methods

---

You are so right. I have forsaken one of the core principles.